Fix filesystem race in bridging PCH emission/loading

dexonsmith opened the pull request:
This fixes a filesystem race in bridging PCH emission/loading, as part of adapting to a more precise in-memory module cache in:
https://github.com/apple/swift-clang/pull/272

These pull-requests need to be tested together.

rdar://problem/48740787

dexonsmith committed:
ClangImporter: Adjust to r355777, PCMCache => ModuleCache

Upstream Clang renamed getPCMCache() to getModuleCache() in r355777,
stopped passing it the Preprocessor passing it directly to ASTReader
instead.  Adapt to those API changes.

rdar://problem/48545366
(cherry picked from commit 131cff72b2044d3a85e79bd539cf8868a4cd1241)

dexonsmith committed:
ClangImporter: Share a module cache between CompilerInstances

Share an in-memory module cache between Clang compiler instances to
maintain a consistent view of the implicitly-generated PCMs, protecting
against filesystem races.  This cache is more robust after r355778.

Note that there are still potential races for PCHs; I'll fix that in a
follow-up.

rdar://problem/48545366
(cherry picked from commit bf0e9a70c6ecaf474d6da55af2bdb8bd1a5b5bcc)

dexonsmith committed:
ClangImporter: Use the new CacheGeneratedPCH in Clang

This flag causes the generated PCH to be written into the in-memory
module cache.  Every CompilerInstance that tries to read the PCH will
use it rather than reading a potentially-since-overwritten version from
the filesystem.

rdar://problem/48740787
(cherry picked from commit 236fc191211e7bf70ab8fbf5639fa1c8614b5336)

dexonsmith commented:
On the swift PR:
Please test with following pull request:
https://github.com/apple/swift-clang/pull/272

@swift-ci Please test

dexonsmith commented:
@jrose-apple is this the right base branch, or should I rebase onto (and merge to) master?

jrose-apple commented:
Master, please. All new content goes to master, then gets cherry-picked where necessary. (In this case it's probably not necessary because [5.1 is getting rebranched one more time](https://swift.org/blog/5-1-release-process/).)

jrose-apple commented:
Oh, since signatures change I guess it needs to go to both branches.

dexonsmith commented:
> Oh, since signatures change I guess it needs to go to both branches.

Yeah, I was hoping committing to the release branch would auto-merge to the master branch.  But I've split off a separate PR for master:
https://github.com/apple/swift/pull/23248

jrose-apple commented:
Is there a way to write a test for this?

dexonsmith commented:
> Is there a way to write a test for this?

I'm not sure.  On the Clang side there's a test that that lang option does the right thing (in https://github.com/apple/swift-clang/pull/272).  In ClangImporter, do you have a way of testing how you invoke Clang?  Or can we do something similar with a unit test?

Note that I think this would be really hard to test in lit.

jrose-apple commented:
> In ClangImporter, do you have a way of testing how you invoke Clang?

I'm not really sure what you mean by this. We can print the invocation that gets constructed but that won't actually tell us anything about whether the PCH is loaded from memory or not.

> Or can we do something similar with a unit test?

This one seems possible, since there's a way to explicitly ask for the PCH to be generated if it doesn't exist yet.

dexonsmith committed:
ClangImporter: Test that PCHs are actually cached in memory

Follow up 236fc191211e7bf70ab8fbf5639fa1c8614b5336 with a ClangImporter
test that we get the PCH written by this process even if it has been
replaced on disk.

dexonsmith commented:
>> Or can we do something similar with a unit test?
>
> This one seems possible, since there's a way to explicitly ask for the PCH to be generated if it doesn't exist yet.

Yup, this worked.  Test added.

dexonsmith commented:
Please test with following pull request:
apple/swift-clang#272

@swift-ci Please test

dexonsmith commented:
@jrose-apple okay to merge this once the tests come back?

jrose-apple commented:
[Today is 5.1 rebranch day](https://swift.org/blog/5-1-release-process/), so it's probably best to merge on master and then wait to see if your change makes it into the rebranch anyway. Or coordinate with Mishal.

dexonsmith commented:
> [Today is 5.1 rebranch day](https://swift.org/blog/5-1-release-process/), so it's probably best to merge on master and then wait to see if your change makes it into the rebranch anyway. Or coordinate with Mishal.

@shahmishal, what do you want to do here?  I just rebased and pushed https://github.com/apple/swift-clang/pull/272 and merged https://github.com/apple/swift/pull/23248.

shahmishal commented:
@dexonsmith I will re-branch swift-5.1-branch from master at 1pm PST today.

shahmishal commented:
@dexonsmith We re-branched from master, we should close this PR.

jrose-apple commented:
The master PR got reverted, so I think this is still relevant. (You can tell by the diff.)

dexonsmith commented:
This landed in 43e771e55b940a0277bef8fe7c58f93a10b25dbb on master and 50a69b01dc82cf7ad8c229e15a1ca6cd2e371e32 on swift-5.1-branch.
